primitives/utility/src/lib.rs

```diff
@@ -51,3 +58,199 @@ impl<T: UpwardMessageSender, W: WrapVersion> SendXcm for ParentAsUmp<T, W> {
+		if let Some(AssetTraderRefunder {
+			mut weight_outstanding,
+			outstanding_concrete_asset: MultiAsset { id, fun },
+		}) = self.0.clone()
+		{
+			let weight = weight.min(weight_outstanding);
+
+			// Get the local asset id in which we can refund fees
+			let (local_asset_id, outstanding_balance) =
+				Matcher::matches_fungibles(&(id.clone(), fun).into()).ok()?;
+
+			// Calculate asset_balance
+			// This read should have already be cached in buy_weight
+			let asset_balance: u128 =
+				FeeCharger::charge_weight_in_fungibles(local_asset_id, weight)
+					.ok()?
+					.saturated_into();
+
+			// Convert into u128 outstanding_balance
+			let outstanding_balance: u128 = outstanding_balance.saturated_into();
+
+			// Construct outstanding_concrete_asset with the same location id and substracted balance
+			let outstanding_concrete_asset: MultiAsset =
+				(id.clone(), (outstanding_balance.saturating_sub(asset_balance))).into();
+
+			// Substract from existing weight and balance
+			weight_outstanding = weight_outstanding.saturating_sub(weight);
+
+			// Override AssetTraderRefunder
+			self.0 = Some(AssetTraderRefunder { weight_outstanding, outstanding_concrete_asset });
+
+			// Only refund if positive
+			if asset_balance > 0 {
+				Some((id, asset_balance).into())
+			} else {
+				None
+			}
+		} else {
+			None
+		}
+	}
+}
+
+impl<
+		AccountId,
+		FeeCharger: ChargeWeightInFungibles<AccountId, ConcreteAssets>,
+		Matcher: MatchesFungibles<ConcreteAssets::AssetId, ConcreteAssets::Balance>,
+		ConcreteAssets: fungibles::Mutate<AccountId>
+			+ fungibles::Transfer<AccountId>
+			+ fungibles::Balanced<AccountId>,
+		HandleRefund: TakeRevenue,
+	> Drop for TakeFirstAssetTrader<AccountId, FeeCharger, Matcher, ConcreteAssets, HandleRefund>
+{
+	fn drop(&mut self) {
+		if let Some(asset_trader) = self.0.clone() {
+			HandleRefund::take_revenue(asset_trader.outstanding_concrete_asset);
+		}
+	}
+}
+
+/// XCM fee depositor to which we implement the TakeRevenue trait
+/// It receives a fungibles::Mutate implemented argument, a matcher to convert MultiAsset into
+/// AssetId and amount, and the fee receiver account
+pub struct XcmFeesToAccount<ConcreteAssets, Matcher, AccountId, ReceiverAccount>(
+	PhantomData<(ConcreteAssets, Matcher, AccountId, ReceiverAccount)>,
+);
+impl<
+		ConcreteAssets: fungibles::Mutate<AccountId>,
+		Matcher: MatchesFungibles<ConcreteAssets::AssetId, ConcreteAssets::Balance>,
+		AccountId: Clone,
+		ReceiverAccount: frame_support::traits::Get<Option<AccountId>>,
+	> TakeRevenue for XcmFeesToAccount<ConcreteAssets, Matcher, AccountId, ReceiverAccount>
+{
+	fn take_revenue(revenue: MultiAsset) {
+		match Matcher::matches_fungibles(&revenue) {
+			Ok((asset_id, amount)) =>
+				if let Some(receiver) = ReceiverAccount::get() {
+					if !amount.is_zero() {
+						let ok = ConcreteAssets::mint_into(asset_id, &receiver, amount).is_ok();
```

Comment:
I'm not 100% sure about this mint, why can't we instead transfer the balance to the recipient account?

Reply:
Because `WIthdrawAsset` burns the assets according to https://github.com/paritytech/polkadot/blob/f6831e91ab574d9e33bf75e9368e8f86a08b8dcb/xcm/xcm-builder/src/fungibles_adapter.rs#L252. Hence there is a "burnt" number of tokens we need to "mint" in the receiver account

Reply:
But isn't this a leaky abstraction? Minting something here assumes that we're using the fungibles adapter to handle assets, which might not always be the case.

If we need to make use of the entity that is used to withdraw assets from to deposit them back into an account, then I think the best idea here is to accept a `TransactAsset` type parameter to this struct, and make use of the `deposit_asset` method there.

It would also be a good idea to mention in the doc comments that the same `TransactAsset` type that is used to withdraw the fungible asset must be supplied here.

Reply:
Oh I see, yeah makes sense. Let me modify this